### PR TITLE
augmentation module + random horizontal flip

### PR DESCRIPTION
**Pull request by dvd42**
addresses part of #298

### CODE REVIEW

File: kornia/augmentation/augmentations.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    def __init__(self, p=0.5):
+        super(RandomHorizontalFlip, self).__init__()
+        self.p: float = p
+
+    def __repr__(self):
+        repr = "(p={0})".format(self.p)
+        return self.__class__.__name__ + repr
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:  # type: ignore
```
Comment (ducha-aiki) [inline kornia/augmentation/augmentations.py]: -> Tuple[torch.Tensor, torch.Tensor]

File: kornia/augmentation/augmentations.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        return random_hflip(input, self.p)
+
+
+def random_hflip(input, p=0.5):
```
Comment (ducha-aiki) [inline kornia/augmentation/augmentations.py]: Typing is missing

File: test/augmentation/test_augmentation.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+        expected = expected.repeat(2, 1, 1)
+
+        actual = op_trace(input)
+
+        assert_allclose(actual, expected)
+
+    def test_gradcheck(self):
+
+        input = torch.tensor([[0., 0., 0.],
```
Comment (ducha-aiki) [inline test/augmentation/test_augmentation.py]: Wouldn`t it better to use random for gradcheck, as we do typically?
E.g. https://github.com/kornia/kornia/blob/master/test/geometry/test_linalg.py#L76

File: kornia/__init__.py [hunk 4]
```diff
@@ -70,5 +71,10 @@
     save_pointcloud_ply,
     load_pointcloud_ply,
 )
+from kornia.augmentation import (
+    random_hflip,
+    RandomHorizontalFlip,
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/__init__.py]: Why do we want to expose this to the top level?

**Comment by ducha-aiki** [comment]
LGTM

File: kornia/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/__init__.py]: that's something to discuss. The policy I was following is to expose to top level all the functionals and keep modules in their own workspace. I like having all in top level just for commodity but whatever is decided it's fine to me.

File: kornia/augmentation/augmentations.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import torch
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: @dvd42 for next PRs imports order: system, torch, kornia ;D

File: kornia/augmentation/augmentations.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+class RandomHorizontalFlip(nn.Module):
+
+    r"""Horizontally flip a tensor image or a batch of tensor images randomly with a given probability.
+    Input should be a tensor of shape (C, H, W) or a batch of tensors :math:`(*, C, H, W)`.
+
+    Args:
+        p (float): probability of the image being flipped. Default value is 0.5
+    """
+
+    def __init__(self, p=0.5):
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: typing

File: kornia/augmentation/augmentations.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        super(RandomHorizontalFlip, self).__init__()
+        self.p: float = p
+
+    def __repr__(self):
+        repr = "(p={0})".format(self.p)
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: typing, fstring

File: kornia/augmentation/augmentations.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        return self.__class__.__name__ + repr
+
+    def forward(self, input: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:  # type: ignore
+        return random_hflip(input, self.p)
+
+
+def random_hflip(input, p=0.5) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: signature proposal as discussed in https://github.com/kornia/kornia/issues/298
```python
UnionTensor =  Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]

def random_hflip(input, p: float, return_transformation: bool = False) -> UnionTensor:
```

File: kornia/augmentation/augmentations.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    r"""Horizontally flip a tensor image or a batch of tensor images randomly with a given probability.
+    Input should be a tensor of shape (C, H, W) or a batch of tensors :math:`(*, C, H, W)`.
+
+    Args:
+        p (float): probability of the image being flipped. Default value is 0.5
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: wouldn't give a default value

File: kornia/augmentation/augmentations.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    Returns:
+        torch.Tensor: The horizontally flipped input
+        torch.Tensor: The applied transformation matrix
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: update the description explaining that depending on the returning flag can be tuple or single raw tensor

File: kornia/augmentation/augmentations.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+    if not torch.is_tensor(input):
+        raise TypeError(f"Input type is not a torch.Tensor. Got {type(input)}")
+
+    if not isinstance(p, float):
+        raise TypeError(f"The probability should be a float number.")
+
+    device: torch.device = input.device
+
+    if len(input.shape) > 3:
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: to handle 3D tensors I would just unsqueeze(0) here and squeeze(0) at very end

File: kornia/augmentation/augmentations.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        probs: torch.Tensor = torch.empty(input.shape[0], device=device).uniform_(0, 1)
+    else:
+        input = input.unsqueeze(0)
+        probs = torch.empty(1, device=device).uniform_(0, 1)
+
+    to_flip: torch.Tensor = probs < p
+
+    if input[to_flip].nelement() != 0:
+
+        flipped: torch.Tensor = hflip(input[to_flip])
```
Comment (edgarriba) [inline kornia/augmentation/augmentations.py]: I don't think this is correct. You select only the tensors to flip and return them, what happen to the others ? I expect to have in a batch flipped and non flipped tensors.

I propose this:
```python
# flip data
flipped = input.clone()
flipped[to_flip] = hflip(input[to_flip])

if return_transformation:
  # transform matrix
  batch_size: int = input.shape[0]
  width: int = input.shape[-1]
  flip_mat = torch.tensor([
    [-1., 0., width],
    [0., 1., 0.],
    [0., 0., 0.]]
  )
  trans_mat =  torch.eye(3, device=input.device, dtype=input.dtype).expand(batch_size, -1, -1)
  trans_mat[to_flip] = flip_mat.to(input.device).to(input.dtype)
  return flipped, trans_mat

return flipped

```

File: kornia/augmentation/augmentations.py (hunk 9 above)
Comment (dvd42) [reply kornia/augmentation/augmentations.py]: Any reason why probability should not be an optional parameter?

File: kornia/augmentation/augmentations.py (hunk 12 above)
Comment (dvd42) [reply kornia/augmentation/augmentations.py]: You are right my intention was to return a batch with the flipped and non-flipped tensors, this is a clear bug. Good thing you spotted it!. Will updata tehe PR with the changes soon

File: kornia/augmentation/augmentations.py (hunk 9 above)
Comment (edgarriba) [reply kornia/augmentation/augmentations.py]: not a fan of default values, but whatever it makes it easier. I see torchvision uses 0.5 by default, let's keep it.

File: kornia/augmentation/augmentations.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    flipped: torch.Tensor = input.clone()
+
+    trans_mat: torch.Tensor = torch.eye(3, device=device, dtype=dtype).expand(input.shape[0], -1, -1)
+
+    flipped[to_flip] = hflip(input[to_flip])
+    flipped.squeeze_()
+
+    if return_transformation:
+
+        w: int = input.shape[-2]
```
Comment (kshitij12345) [inline kornia/augmentation/augmentations.py]: This should be `w= input.shape[-1] `, right? Since the Tensor Dimensions are `*, C, H, W or C, H, W`.

Comment (edgarriba) [reply kornia/augmentation/augmentations.py]: Good catch. Will fix It

**Pull request merged**